include/rcpputils/patch.hpp

```diff
@@ -0,0 +1,200 @@
+};
+
+template<size_t N, typename SignatureT>
+struct trampoline;
+
+template<size_t N, typename ReturnT, typename ... ArgTs>
+struct trampoline<N, ReturnT(ArgTs...)>
+{
+  static ReturnT base(ArgTs... args)
+  {
+    return target(std::forward<ArgTs>(args)...);
+  }
+
+  static std::function<ReturnT(ArgTs...)> target;
+};
+
+template<size_t N, typename ReturnT, typename ... ArgTs>
+std::function<ReturnT(ArgTs...)>
+trampoline<N, ReturnT(ArgTs...)>::target;
+
+template<size_t N, typename SignatureT>
+auto prepare_trampoline(std::function<SignatureT> target)
+{
+  trampoline<N, SignatureT>::target = target;
+  return trampoline<N, SignatureT>::base;
+}
+
+template<size_t N, typename SignatureT>
+class patch;
+
+template<size_t N, typename ReturnT, typename ... ArgTs>
+class patch<N, ReturnT(ArgTs...)>
+{
+public:
+  using mock_type = typename patch_traits<N, ReturnT(ArgTs...)>::mock_type;
+
+  patch(const std::string & target, std::function<ReturnT(ArgTs...)> proxy)
+  : proxy_(proxy)
+  {
+    auto MMK_MANGLE(mock_type, create) =
+      patch_traits<N, ReturnT(ArgTs...)>::MMK_MANGLE(mock_type, create);
+    mock_ = mmk_mock(target.c_str(), mock_type);
+  }
+
+  patch(const patch &) = delete;
+  patch & operator=(const patch &) = delete;
+
+  patch(patch && other)
+  {
+    mock_ = other.mock_;
+    other.mock_ = nullptr;
+  }
+
+  patch & operator=(patch && other)
+  {
+    if (mock_) {
+      mmk_reset(mock_);
+    }
+    mock_ = other.mock_;
+    other.mock_ = nullptr;
+  }
+
+  ~patch()
+  {
+    if (mock_) {
+      mmk_reset(mock_);
+    }
+  }
+
+  patch & then_call(std::function<ReturnT(ArgTs...)> replacement) &
+  {
+    auto type_erased_trampoline =
+      reinterpret_cast<mmk_fn>(prepare_trampoline<N>(replacement));
+    mmk_when(proxy_(any<ArgTs>()...), .then_call = type_erased_trampoline);
+    return *this;
+  }
+
+  patch && then_call(std::function<ReturnT(ArgTs...)> replacement) &&
+  {
+    auto type_erased_trampoline =
+      reinterpret_cast<mmk_fn>(prepare_trampoline<N>(replacement));
+    mmk_when(proxy_(any<ArgTs>()...), .then_call = type_erased_trampoline);
+    return std::move(*this);
+  }
+
+private:
+  template<typename T>
+  T any() {return mmk_any(T);}
+
+  mock_type mock_;
+  std::function<ReturnT(ArgTs...)> proxy_;
+};
+
+template<size_t N, typename SignatureT>
+auto make_patch(const std::string & target, std::function<SignatureT> proxy)
+{
+  return patch<N, SignatureT>(target, proxy);
+}
+
+}  // namespace rcpputils
+
+#define RCPPUTILS_PATCH_IGNORE_OPERATOR(type, op) \
+  bool operator op(const type &, const type &) { \
+    return false; \
+  }
+
+#define RCPPUTILS_PATCH_PROXY(func) \
+  [] (auto && ... args)->decltype(auto) { \
+    return func(std::forward<decltype(args)>(args)...); \
+  }
+
+#define RCPPUTILS_PATCH_TARGET(what, where) \
+  (std::string(RCUTILS_STRINGIFY(where)) + "@" + (what))
+
+#define patch(what, where, with) \
```

Comment:
Your macro name is the same as your struct name, will that lead to confusion?

Reply:
Perhaps. I don't expect the base class being used directly, but we can rename it.